### ISSUE TITLE
Add support for software breakpoints

### ISSUE THREAD

**Issue by Wenzel** [Wenzel/r2vmi#18]
Software breakpoints are not stealthly, but they are much faster than our memory access events tricks.

current behavior:

![bp](https://user-images.githubusercontent.com/964610/45269757-e7ff5100-b49b-11e8-996c-a1d9d99a5451.png)

message if from LibVMI
`VMI_ERROR: process_interrupt_event Need to specify reinjection behaviour!`

Upstream issue is opened:
https://github.com/libvmi/libvmi/issues/660
EDIT: issue has been closed.

**Comment by Wenzel** [comment]
TODO:

- [x] Is `radare2` responsible for writing the `0xCC` into memory ? Or should i do it {myself](https://github.com/Wenzel/r2vmi/blob/master/debug_vmi.c#L584) ?
- [ ] Am i supposed to see the `int 3` in the disassembly output ?
`r2 -d /bin/ls`
![screenshot at 2018-09-11 16-59-08](https://user-images.githubusercontent.com/964610/45364953-16884380-b5e4-11e8-98ca-e0fbb38de220.png)
`r2 -d vmi://vm:process`
![screenshot at 2018-09-12 04-51-20](https://user-images.githubusercontent.com/964610/45397199-a01d2d00-b647-11e8-9001-d5232bf4a4ce.png)

- [ ] how to write the original opcode back into memory before singlestepping ?

**Comment by Wenzel** [comment]
From https://github.com/radare/radare2/blob/b25653ec2ae48bad9a554753d0c21ce86b0e7e0a/libr/debug/p/debug_native.c#L1523:
~~~C
/*
 * set or unset breakpoints...
 *
 * we only handle the case for hardware breakpoints here. otherwise,
 * we let the caller handle the work.
 */
~~~

So if `__breakpoint` returns false, the caller will handle the software breakpoints ?

**Comment by Wenzel** [comment]
Related: #22

**Comment by Wenzel** [comment]
After refactoring the code and letting `radare2` manage the software breakpoints, i can still see the `int3` instead of the original opcode:
![int3](https://user-images.githubusercontent.com/964610/45497513-af53c600-b780-11e8-9b91-7c671feb4961.png)

**Comment by Wenzel** [comment]
- [ ] when `radare2` sets software breakpoint with the native debugger, you can call continue without hitting the same breakpoint again.
- [ ] i'm not supposed to see the `int3` in the disassembly. related to first issue i guess

**Comment by Wenzel** [comment]
Fixed by #27 :tada:
